### ISSUE TITLE
RFE: Alternate cryptographic backends

### ISSUE THREAD

**Issue by DemiMarie** [ctz/rustls#521]
rustls currently depends on *ring* as the only crypto backend.

While *ring* is certainly a good crypto library, it has a poor story regarding backwards compatibility.  Additionally, it is statically linked, which bloats executables.

Therefore, I am filing this RFE to request support for alternate cryptographic implementations, such as OpenSSL.  This would require figuring out a story for certificate verification, since we don’t want to use OpenSSL’s certificate APIs and WebPKI only supports *ring*.  That said, on Windows and macOS there are valid reasons for just using the platform’s native certificate verification APIs instead.

**Comment by djc** [comment]
See also: #390.

**Comment by DemiMarie** [comment]
Another advantage of this approach is that a FIPS 140-2 certified implementation could be used where needed.  I don’t have such a need, but others do.

**Comment by djc** [comment]
What is your issue with *ring*'s approach to backwards compatibility? Have you restricted the number of available ciphersuites so that executable sizes can be reduced at link time?

**Comment by DemiMarie** [comment]
> What is your issue with _ring_'s approach to backwards compatibility?

Updates to _ring_ require time before they can propagate through the ecosystem; during this time, people are running an unsupported version.  If _ring_ provided security support to major version N - 1, this would not be anywhere near as much of a problem.  Hopefully, _ring_ will reach 1.0 and its API will stabilize.

> Have you restricted the number of available ciphersuites so that executable sizes can be reduced at link time?

That isn’t always practical for compatibility reasons, and it still won’t be able to beat linking to a shared system library such as OpenSSL.  The vast majority of nasty OpenSSL CVEs have been in the TLS and ASN.1 code, not the cryptography, and I suspect many of the cryptographic ones affect _ring_ too.  Using a system crypto library means that I can rely on the OS to handle security updates for it, and that those who need FIPS 140-2 compliance can use rustls at all.

**Comment by Darkspirit** [comment]
https://github.com/briansmith/ring/issues/535 seems to be planned for the upcoming 0.17 release: https://github.com/briansmith/ring/pull/849#issuecomment-742863466

Could https://github.com/abetterinternet/crustls become a sufficient system package?
Otherwise it would be great if https://github.com/mesalock-linux/mesalink could become one. An OpenSSL 3.0 API might make it even more attractive for the upcoming transition. 

DTLS support is tracked in https://github.com/ctz/rustls/issues/40. ([DTLS 1.3](https://datatracker.ietf.org/doc/draft-ietf-tls-dtls13/) IESG state: "In Last Call (ends 2021-02-22)")

As a community we should aim for getting rustls + webpki + ring certified.

**Comment by DemiMarie** [comment]
> As a community we should aim for getting rustls + webpki + ring certified.

FIPS 140-2 certification is very expensive (tens of thousands of USD, IIRC) and very time-consuming (months or years).  Only _ring_ would need to be certified, since that’s where the actual crypto is done, but I still believe that trying to get _ring_ FIPS-certified would be a mistake.  LibreSSL _dropped_ FIPS support, and Go never supported it at all.  There is a branch of Go using the boringcrypto library (based on BoringSSL) that provides an alternative for those users.

In short, the best approach to FIPS 140-2 compliance that I am aware of is to simply offload all cryptography to an OS-provided library.  Users who need FIPS 140-2 compliance will then merely need to choose an OS that is already FIPS 140-2 certified, which they would likely need to do anyway.

**Comment by briansmith** [comment]
@DemiMarie Do you yourself need FIPS support? If not, why don't we drop that topic and wait for somebody who actually needs it to ask for it.

**Comment by briansmith** [comment]
BTW, tens of thousands of dollars isn't a lot of money for these projects. Hundreds of thousands of dollars have already been invested into them.

**Comment by DemiMarie** [comment]
@briansmith I do not.  I merely brought it up because I know some people do, and as a reason someone might want to use an alternate crypto backend.

**Comment by DemiMarie** [comment]
> @DemiMarie Do you yourself need FIPS support? If not, why don't we drop that topic and wait for somebody who actually needs it to ask for it.

Red Hat Enterprise Linux needs it.  My understanding is that rustls cannot ship in RHEL unless it uses one of RHEL’s existing cryptographic libraries.  IIRC, those are libgcrypt, Nettle, NSS, and OpenSSL, along with the kernel cryptographic code.

**Comment by sehz** [comment]
Ran into so many problem with building `ring` (cross-compiling, linking, etc).   This is one of the reason why we dropped `rustls` in favor of `openssl`.   Seems like `crypto` should be treated like OS service.

**Comment by DemiMarie** [comment]
> Ran into so many problem with building `ring` (cross-compiling, linking, etc). This is one of the reason why we dropped `rustls` in favor of `openssl`. Seems like `crypto` should be treated like OS service.

It is in fact treated as such, on all major platforms.

@ctz: would you accept a high-quality PR that included an OpenSSL backend?

**Comment by Ralith** [comment]
People who need that might be better served by using rust-native-tls.

**Comment by DemiMarie** [comment]
> People who need that might be better served by using rust-native-tls.

I consider rustls to be a better TLS implementation than OpenSSL.  The majority of OpenSSL’s security holes have not been in the crypto, but rather memory safety issues elsewhere.  Since most of the problems with rustls seem to be due to _ring_, using OpenSSL for the crypto instead seems to be an obvious alternative.

**Comment by TotalKrill** [comment]
I have also encountered a lot of issues regarding ring, cross-compiling, linking and the multi version issue of ring ( the most annoying one for me personally ). 

Given the above, I still do not believe that changing rustls to include openssl as a backend is the right approach, since it might actually hinder development for this project. 

I think the correct approach would be to create a wrapper-crate towards the entire rustls, and unyfing on an API, that could in turn switch between rust-native-tls and rustls, preferably in a drop-in manner, meaning that if one were to write an application or library using rustls, but needing a feature that is currently not in rustls because of some dependency. Then that would be treated as a "quick and dirty fix" until such time as it is supported by rustls. 

The "quick and dirty fix" and "until", is the basis for my reasoning, since I believe that creating a drop-in wrapper crate would be faster. And that it should be a treated as a temporary ( yeah, i know) solution to using rustls.

**Comment by briansmith** [comment]
The multi-version issue in *ring* is mostly fixed in *ring*'s main branch now; see https://github.com/briansmith/ring/issues/1268. Review and testing of that by others would be appreciated.

Cross-compilation of *ring* has made big progress in *ring*'s main branch as well. IIRC, there is now a C or Rust fallback for everything, so platforms for which *ring* doesn't have assembly language code are now supported. There is work underway to verify this with the MIPS little-endian port. After that, I think we'll have to clean up some stuff that's not supported for big-endian archs yet, but that should be relatively easy.

**Comment by DemiMarie** [comment]
> The multi-version issue in _ring_ is mostly fixed in _ring_'s main branch now; see [briansmith/ring#1268](https://github.com/briansmith/ring/issues/1268). Review and testing of that by others would be appreciated.
> 
> Cross-compilation of _ring_ has made big progress in _ring_'s main branch as well. IIRC, there is now a C or Rust fallback for everything, so platforms for which _ring_ doesn't have assembly language code are now supported. There is work underway to verify this with the MIPS little-endian port. After that, I think we'll have to clean up some stuff that's not supported for big-endian archs yet, but that should be relatively easy.

Personally, I wold still like to be able to link to system-provided cryptographic libraries instead.  The main reasons (in no particular order) are:

1. It is a hard requirement for packaging in some distributions, such as RHEL and its derivatives.
2. Having to statically link *ring* increases the size of the executable.
3. System-provided cryptographic libraries receive security updates without me having to take any action at all, whereas programs using *ring* need to be recompiled if there is a security issue in *ring*.

In short, I am not particularly interested in `rust-native-tls`, as I consider `rustls` to be better than native platform implementations.  I am much less convinced when it comes to _ring_, and there are significant advantages to using system libraries for actual cryptographic primitives.  On Linux and \*BSD, these problems could be mitigated if _ring_ were a system library I could dynamically link to.

**Comment by djc** [comment]
I now think it would be valuable to add high-level traits in rustls that abstract over the crypto operations. If we were to do that, I don't think we should include other implementations as part of rustls -- it would be up to downstream users to decide whether to trust another implementation. We might even guard the use of alternative implementations with a feature flag like `dangerous_configuration`. I also think that we should likely still restrict alternative implementations to the cipher suites/signature schemes/key exchange groups currently supported in this crate, at least as a first iteration.

**Comment by franziskuskiefer** [comment]
I'd second that. One of the recommendations in the audit was to use formally verified crypto. High-level crypto traits would allow rustls to do so, for example using evercrypt.

**Comment by DemiMarie** [comment]
This is also a hard requirement to use hardware cryptographic accelerators.  These require cryptography to be asynchronous, as they have very deep pipelines that need to be kept full.

**Comment by sonder-joker** [comment]
Is there any further discussion? Our community hopes to be able to provide a backend for wasi-crypto, and a middle layer is essential. I wish to get some tips about how to design it and pull request back to the upstream. https://github.com/WasmEdge/WasmEdge/issues/1430

**Comment by djc** [comment]
There has not been any further discussion. If you or your community wants to work on this, I'd be open to reviewing a potential design. I think essentially it would have to be a high-level set of traits that essentially mimic the current *ring* API to cover the rustls crypto requirements that are not currently covered by traits.

**Comment by dheater** [comment]
Maybe the `tls-api` crate could used, if not directly, then as a starting point for trait/API design: https://docs.rs/tls-api/0.9.0/tls_api/

**Comment by djc** [comment]
That is a very different level of abstraction, which wouldn't allow re-use of the non-cryptography code in rustls.

**Comment by quininer** [comment]
Some time ago, I studied [the API of using pure Rust crate to implement ring 0.16](https://github.com/quininer/ziwaan) (now it's mainly the RustCrypto ecosystem, thanks to the work of RustCrypto team) . The original intention was to make crates such as rustls compile more smoothly under riscv.

In process of implementation, I realized again that Rust ecosystem is not enough to completely replace ring.
for example, the rsa code is not that reliable, p384 has not had a proper code until recently (thank @jedisct1), the parser of formats such as der is not that mature. and it is difficult to compare with ring in performance.

but if we only focus on algorithms like curve25519, chacha20-poly1305, then Rust ecosystem is enough (It can even be done better, for example using [hacspec](https://github.com/hacspec/hacspec/)).

So when designing a alternate backend API, it would be nice to also be able to configure ciphersuite. 
it also brings another benefit, allowing the use of custom cipersuite, such as support non-standard post-quantum algorithms without fork code.

**Comment by tarcieri** [comment]
> the parser of formats such as der is not that mature. 

I'm curious why you say this. We have one of the most sophisticated and complete DER parsers available in Rust, with support for features like decoding/encoding DER from/to streaming (and constant time) PEM decoders/encoders, along with some of the most complete parsers for formats like PKCS#1, PKCS#8, and SEC1.

**Comment by quininer** [comment]
> I'm curious why you say this.

I may have worded something wrong. the der crate is fully functional, but I ran into a panic when testing with der 0.5 (It was fixed in der 0.6 https://github.com/RustCrypto/formats/pull/447).

I'm sure any issues can be fixed, but it may take some time to stabilize the code and API. anyway, this is a bit offtopic.

> and ideally ones designed for performance

You're right, I mention hacspec here for correctness not performance.

**Comment by TieWay59** [comment]
> I think essentially it would have to be a high-level set of traits that essentially mimic the current ring API to cover the rustls crypto requirements that are not currently covered by traits. @djc via [src](https://github.com/rustls/rustls/issues/521#issuecomment-1123343171)

Hi, I'm new here and trying to follow up on the work.

Could someone please enlighten me about _a high-level set of traits_, and what it should be like?

And if there are some crypto requirements not currently covered by traits, where I can find a list or some document of them?

Any helpful information will be appreciated.

I have emailed Dirkjan because I'm not sure about my insight into the discussion above. But I have waited a week and I guess I picked up the wrong email address.

---

In my original imagination, we can swap the Ring types with traits and implement them with different lib support.

Is it going to be like something as below? feel free to correct me, I'll appreciate it.


```mermaid
graph TD
    A(Rustls) --> |replace ring invocation| B(a high-level set of traits)
    B -->|Impl using| D(Ring)
    B -->|Impl using| E(WASI-Crypto)
    B -->|Impl using| F(Other Rust Crypto Infra ...)
```

**Comment by djc** [comment]
In case anyone missed it, there is now a PR with a draft RFC for this work in #1184.

**Comment by dignifiedquire** [comment]
FYI I have a WIP of extracting all usages of the crypto backend details here: https://github.com/dignifiedquire/rustls/tree/feat-crypto-backend 

I don't know when I will have time to work on the implementation of the actual alternative backend, but maybe helps at least in the planning effort :)

**Comment by jsha** [comment]
This is great. Another use case for alternate backends: rustls-ffi uses Miri for testing, but has to disable Miri on most of the interesting code paths because they call out to FFI (assembly) code in *ring*, and Miri does not support calling out to FFI. If we have a backend with no FFI, rustls-ffi should be able to make more extensive use of Miri to verify correctness of its unsafe code.

(note: you might think "rustls-ffi offers an FFI interface; isn't that a problem for Miri?" but the unittests call rustls-ffi code from Rust, so everything is hunky-dory)

**Comment by RalfJung** [comment]
> This is great. Another use case for alternate backends: rustls-ffi uses Miri for testing, but has to disable Miri on most of the interesting code paths because they call out to FFI (assembly) code in ring, and Miri does not support calling out to FFI. If we have a backend with no FFI, rustls-ffi should be able to make more extensive use of Miri to verify correctness of its unsafe code.

Above it was said that ring has fallback paths in Rust -- do those not get used with Miri for some reason or are there no fallback paths for some operations?

(Also obligatory reminder that Miri be default provides deterministic execution, so don't run any code that needs proper randomness in Miri on real data!)

**Comment by jsha** [comment]
Honestly I had missed that there were fallbacks. Though note:

> there is now a C or Rust fallback for everything

A C fallback isn't much use for Miri because it's still FFI, right?

**Comment by dignifiedquire** [comment]
@djc  any updates on this?

**Comment by RalfJung** [comment]
> A C fallback isn't much use for Miri because it's still FFI, right?

True.

**Comment by djc** [comment]
For the current state, have a look at #1405. In short, the next semver-incompatible rustls release (which we expect to happen within a few weeks from now) will have support for pluggable crypto providers.

**Comment by dignifiedquire** [comment]
Amazing, I totally missed that! looking forward to trying it out

**Comment by trevor-vaughan** [comment]
This is awesome progress. I have a (possibly naive) question.

How do we audit what provider the compiled application is configured to use?

As in, given binary X, what do I run to determine if it is using a provider that is allowed by my organization?

**Comment by djc** [comment]
AFAIK https://github.com/rust-secure-code/cargo-auditable could help you tell which dependencies are compiled into your binary. That should be an important part of the puzzle, although in any given binary there might be multiple sites that instantiate a rustls config and all of these might use a different crypto provider, and that might be harder to check from a binary.

**Comment by trevor-vaughan** [comment]
> AFAIK https://github.com/rust-secure-code/cargo-auditable could help you tell which dependencies are compiled into your binary. That should be an important part of the puzzle, although in any given binary there might be multiple sites that instantiate a rustls config and all of these might use a different crypto provider, and that might be harder to check from a binary.

@djc That's what I suspected. It's definitely a use case that should be accounted for at some point but it sounds like the "right answer" is to always compile from scratch at this point. ( _cries in 90's pre-packaging hell_ )

**Comment by DemiMarie** [comment]
> > AFAIK https://github.com/rust-secure-code/cargo-auditable could help you tell which dependencies are compiled into your binary. That should be an important part of the puzzle, although in any given binary there might be multiple sites that instantiate a rustls config and all of these might use a different crypto provider, and that might be harder to check from a binary.
> 
> @djc That's what I suspected. It's definitely a use case that should be accounted for at some point but it sounds like the "right answer" is to always compile from scratch at this point. ( _cries in 90's pre-packaging hell_ )

If you need to only use FIPS 140-certified cryptography, then you will likely need to build everything yourself indeed.  There are lots of other Rust cryptographic libraries out there and you will need to make sure that none of them are built, or replace them with versions that call whatever certified cryptography is on your system.

**Comment by trevor-vaughan** [comment]
@DemiMarie It is certainly making `rust` a minefield for certified crypto and I agree on essentially needing to do a full code audit and build.

Unfortunately, the burden is almost to the point of untenable for large application bases.

**Comment by ctz** [comment]
Could you detail what your requirements are? Is it just the normal FIPS140 ones, ie all cryptography must be approved or allowed, and approved cryptography must be a FIPS140-certified implementation on the hardware and operating environment that certificate was issued for? Anything more than this?

**Comment by trevor-vaughan** [comment]
@ctz That's the one. But that only applies for everything that uses cryptography for the "protection of sensitive data". This means hash algorithms down the stack and, of course, the correct configuration of each.

It gets more complicated by the fact that _technically_ you can use non-FIPS algorithms as long as you're not using them for "protection of sensitive data". It's easier to just ban them outright but not actually required.

So, could I use the [md-5 crate](https://crates.io/crates/md-5) for doing a bunch of checksums to prevent collisions on temp files in the filesystem that can't affect a security path? Sure! Is it hard to determine that I'm using it specifically for that use...yes.

Some languages have the ability to toggle FIPS with a setting that essentially says "I'm doing the bad thing but it's OK". Likewise, some vendor implementations tie them into lower levels that remove the algorithms altogether for safety and break that application path.

But, in the end, the questions that needs to be answered are:

* US and Canada
  * Am I using algorithms that meet the FIPS requirements.
* Other folks that care
  * [ISO/IEC 19790](https://www.iso.org/standard/52906.html) and [ISO/IEC 24759](https://www.iso.org/standard/72515.html) which are the foundation for [FIPS 140-3](https://csrc.nist.gov/Projects/cryptographic-module-validation-program/fips-140-3-standards) (but not sufficient).

**Comment by DemiMarie** [comment]
@trevor-vaughan Do you just need to use approved _algorithms_, or do you need certified _implementations_ of those algorithms?  The latter is likely going to mean sending a bunch of patches upstream — there are good reasons (notably better APIs) why people use pure-Rust crypto implementations over the system libraries.

**Comment by trevor-vaughan** [comment]
@DemiMarie If you need to meet the FIPS requirement, it must be the certified implementation.

Per [the official CMVP page](https://csrc.nist.gov/projects/cryptographic-module-validation-program):

> Non-validated cryptography is viewed by NIST as providing no protection to the information or data—in effect the data would be considered unprotected plaintext. If the agency specifies that the information or data be cryptographically protected, then FIPS 140-2 or FIPS 140-3 is applicable. In essence, if cryptography is required, then it must be validated. Should the cryptographic module be revoked, use of that module is no longer permitted.

**Comment by djc** [comment]
@trevor-vaughan we'd definitely like to make rustls an option for FIPS-certified environments so would be interesting to hear more about what challenges you have with Rust compared to, say, C++. Might also be interesting to bring this up in the wg-secure-code stream in the rustlang Zulip.

**Comment by trevor-vaughan** [comment]
@djc Right now, it's more `rust` vs `golang`. Golang == recompile the thing one of two ways and you're gold (as long as someone didn't do something silly).

In the case of any compiled language, you're definitely at the mercy of the one doing the compiling and I'm _definitely_ starting to have 1990's `zlib` flashbacks :-(.

**Comment by Firstyear** [comment]
The problem we have for OpenSUSE is much simpler - we have to build some projects on arches that *arent* x86_64/aarch64. Ring almost constantly fails to build on these. Because rustls is used so commonly, with no alternatives in projects, then it means we can't build a lot of projects that depend on rustls for other arches. 

So having the ability to set alternate cryptographic backends would help us here by allowing us to build more easily on alternate arches.

**Comment by djc** [comment]
@trevor-vaughan sorry, it's not very clear to me what you mean here with the Go vs Rust compilation issue. I think you're suggesting that Rust's compilation model causes issues for your FIPS story, but I don't understand why?

**Comment by trevor-vaughan** [comment]
@djc I came upon this issue when I was trying to find "the way" to flip the Rust backend over to use `openssl` natively.

Golang provides `crypto` as it's "standard" crypto library and has two methods for swapping it into a FIPS-supported mode. One is to use boringcrypto (and follow the actual FIPS documentation) and the other is provided by Red Hat that essentially shims the underlying `openssl` stack into the crypto hooks and, therefore, inherits the certification from the underlying platform.

IIRC, in the early days, Rust pretty much used `openssl` straight and then decided to essentially modularize the whole crypto set. This made it pretty much impossible to flip any given compile into a FIPS compliant mode since you had to play whack-a-mole with every package out there.

In both cases (and every other language) you do have to make sure that someone isn't dragging in some random package to do crypto but the golang (and javascript, python, etc...) pattern of having a "core" crypto library that 99% of folks use makes it much easier for regulatory focused shops.

**Comment by trevor-vaughan** [comment]
> So from this perspective, would it help you if the rustls _ring_ backend lived in a separate crate, so that you could check your `Cargo.lock` file to make sure that it's not getting compiled in?

🤔 Actually, yes, that might help quite a bit. Any way that makes it easier to determine *exactly* what is in use (or not) is a win.

**Comment by ctz** [comment]
We've made a start on this in the recent 0.22 release. Shipping FIPS140 support is tracked in #1540.

**Comment by TaaviE** [comment]
Has there been any progress in terms of providing a Rust-only backend that does not require a (fragile) cross-compilation stack?

**Comment by tarcieri** [comment]
@TaaviE there's https://github.com/RustCrypto/rustls-rustcrypto which is built on @RustCrypto's pure Rust implementations

**Comment by djc** [comment]
> Has there been any progress in terms of providing a Rust-only backend that does not require a (fragile) cross-compilation stack?

You should have a look at [Graviola](https://docs.rs/graviola/latest/graviola/).

**Comment by TaaviE** [comment]
@tarcieri I saw it, but can't get it to build due to issues with its dependencies.

@djc I'll check it out, thanks!

**Comment by tarcieri** [comment]
@TaaviE yeah, working through stable releases of all the @RustCrypto crates right now, hopefully things should settle down soon

**Comment by newpavlov** [comment]
@TaaviE 
Have you tried the version published on crates.io? It uses somewhat old dependencies, but hopefully we will cut a new release soon.

**Comment by TaaviE** [comment]
> Have you tried the version published on crates.io? It uses somewhat old dependencies, but hopefully we will cut a new release soon.

I tried the latest that was compatible with the `rustc` available to me on the machine, I'll soon try again with a newer version. Graviola however seems to build. So there's certainly hope that there's a cross-compilation option that isn't so cumbersome.
